src/ret_val.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+#ifndef RET_VAL_H
+#define RET_VAL_H
+
+#include "debug.h"
+
+#include <string>
+#include <type_traits>
+
+template<typename T, typename std::enable_if<std::is_same<T, bool>::value, int>::type = 0>
+struct def_success_const : std::integral_constant<bool, true> {};
+
+template<typename T, typename std::enable_if<std::is_same<T, bool>::value, int>::type = 0>
+struct def_failure_const : std::integral_constant<bool, false> {};
+
+template<typename T, T default_success = def_success_const<T>::value, T default_failure = def_failure_const<T>::value>
```

Comment:
The `default_success` and `default_failure` are a bit too much. Just let the user of the class decide what is and isn't a failure. Or even easier: add a `bool was_success;` to the class, which indicates (independently of the "return" value) whether it is considered success.

Currently one could for example call `failure` with a code equal to `default_success`, yielding an object that will report success, even through it was initialized via `failure`. A `was_success` member would enforce that all objects created via `sucess`/`failure` report as such.

Reply:
There's a check that prevents it, but yes, the templates are needlessly complex and aren't flexible enough. I've simplified the specialization (but haven't pushed yet).